src/Gaffer/MetadataAlgo.cpp

```diff
@@ -222,6 +224,89 @@ void bookmarks( const Node *node, std::vector<NodePtr> &bookmarks )
 	}
 }
 
+void setNumericBookmark( Node *node, int value )
+{
+	if( value < 0 )
+	{
+		return;
+	}
+
+	if( readOnly( node ) )
+	{
+		return;
+	}
+
+	// \todo: should this be const (and the `root` arg in nodesWithMetadata as well?)
+	ScriptNode *scriptNode = node->ancestor<ScriptNode>();
+	if( !scriptNode )
+	{
+		return;
+	}
+
+	for( Node *nodeWithMetadata : Metadata::nodesWithMetadata( scriptNode, g_numericBookmarkName, /* instanceOnly = */ true ) )
+	{
+		if( getNumericBookmark( nodeWithMetadata ) != value )
+		{
+			continue;
+		}
+
+		if( readOnly( nodeWithMetadata ) )
```

Comment:
_Here_ I think we do need to be checking for read-onliness, so although I stand by my comment on the check above (line 234), I think something along these lines is kosher here. I wonder if the [approach I took in TransformTool](https://github.com/GafferHQ/gaffer/blob/master/src/GafferSceneUI/TransformTool.cpp#L81-L93) is slightly more appropriate though. We definitely don't want to modify anything inside a reference, which is what `ancestorMakesChildNodesReadOnly()` is essentially checking for. But perhaps we _do_ want to steal the bookmark from a read-only node that isn't in a reference (that has just been locked via the NodeEditor)? Or at least leave it to the UI to decide? I'm not sure - it's a very subtle distinction...